### PR TITLE
fix: measured times for application startup

### PR DESCRIPTION
**Pull request by arkaitzgarro**
Split the startup time in two:
- Application load time, usually taken by Composer, file reading, etc
- Framework load, measured by Laravel's events `booting` and `booted`.

Add the time that it takes for a request to be handled, measured by the time between `RouteMatched` and `RequestHandled` events. I suppose this also includes middlewares.

Add the query name: query type + target tables.

Screenshot 2020-01-24 at 09 50 27

### CODE REVIEW

**Comment by arkaitzgarro** [comment]
@countless-integers im going to clean up the code a bit, and extract parts from the `ServiceProvider` to separate classes and start writing some tests.

File: src/ServiceProvider.php [hunk 1]
```diff
@@ -121,4 +137,28 @@ protected function getAppConfig(): array
 
         return $config;
     }
+
+    protected function getController()
```
Comment (countless-integers) [inline src/ServiceProvider.php]:
```suggestion
    protected function getController(): ?string
```

File: src/Collectors/DBQueryCollector.php [hunk 2]
```diff
@@ -61,4 +57,23 @@ public static function getName(): string
...
         return 'query-collector';
     }
+
+    private function getQueryName(string $sql): string
+    {
+        $fallback = 'Eloquent Query';
+
+        try {
+            $query_type = QuerySplitter::getQueryType($sql);
+            $tables = QuerySplitter::splitTables($sql);
```
Comment (countless-integers) [inline src/Collectors/DBQueryCollector.php:83]: Interesting package, lots of regexes. Builds are failing, but it seems to do what we need it to do.

File: src/ServiceProvider.php [hunk 3]
```diff
@@ -55,14 +57,28 @@ protected function registerAgent(): void
...
+    protected function listenForEvents(): void
     {
+        $start_time = $this->app['request']->server('REQUEST_TIME_FLOAT') ?? microtime(true);
         $timeline_collector = $this->app->make(Agent::class)->getCollector(TimelineDataCollector::getName());
+        
+        // Application and Laravel startup times
+        $timeline_collector->startMeasure('app_boot', 'app', 'boot', 'App boot', $start_time);
+        $this->app->booting(function () use ($timeline_collector) {
+            $timeline_collector->stopMeasure('app_boot');
+            $timeline_collector->startMeasure('laravel_boot', 'laravel', 'boot', 'Laravel boot');
```
Comment (countless-integers) [inline src/ServiceProvider.php]: What's the difference between `app_boot` and `laravel_boot`?

File: src/Collectors/DBQueryCollector.php [hunk 4]
```diff
@@ -61,4 +57,23 @@ public static function getName(): string
...
+
+            if (isset($query_type) && is_array($tables)) {
+                // Query type and tables
+                return $query_type . ' ' . join(', ', array_values($tables));
+            }
+
+            return $fallback;
+        } catch (Exception $e) {
+            return $fallback;
```
Comment (countless-integers) [inline src/Collectors/DBQueryCollector.php:92]: You could also keep the catch block empty and move the `return $fallback` after the whole try-catch block to same 1 line of code (#crazy-gains).

File: src/Collectors/DBQueryCollector.php (hunk 2 above)
Comment (arkaitzgarro) [reply src/Collectors/DBQueryCollector.php:83]: Yeap, but it does the job, and it's also really fast.

File: src/ServiceProvider.php (hunk 3 above)
Comment (arkaitzgarro) [reply src/ServiceProvider.php]: We might need a better naming indeed. If I understand correctly how Laravel works, the difference between both of them is as follows:
- `app_boot`:  the time for loading Composer autoloader file, create Laravel application instance, create the container and register service providers.
- `laravel_boot`: load environment variables, register facades, resolve the Kernel, register routes, register middlewares...

File: src/Collectors/DBQueryCollector.php (hunk 4 above)
Comment (arkaitzgarro) [reply src/Collectors/DBQueryCollector.php:92]: I never liked empty `catch` blocks 😅

Comment (countless-integers) [reply src/Collectors/DBQueryCollector.php:92]: Fair enough.

**Review by countless-integers — APPROVED** [review]
🚢

**Pull request merged**